Source/Urho3D/Scene/PrefabReference.cpp

```diff
@@ -20,76 +20,247 @@
+
+    for (int i = numComponents - 1; i >= 0; --i)
+    {
+        Component* component = components[i];
+        if (component->IsTemporary())
         {
-            //Skip inner prefabs
-            if (child == prefabRef->GetRootNode())
+            if (component != this)
+                node->RemoveComponent(component);
+            else
             {
-                isPrefabRoot = true;
-                break;;
+                URHO3D_LOGWARNING("PrefabReference component should not be temporary");
+                component->SetTemporary(false);
             }
         }
-        if (!isPrefabRoot)
-        {
-            SetTemporaryFlag(child, isTemporary);
-        }
     }
 }
-} // namespace
 
-PrefabReference::PrefabReference(Context* context)
-    : BaseClassName(context)
-    , prefabRef_(XMLFile::GetTypeStatic())
+void PrefabReference::RemoveTemporaryChildren(Node* node) const
 {
+    const auto& children = node->GetChildren();
+    const int numChildren = static_cast<int>(node->GetNumChildren());
+
+    for (int i = numChildren - 1; i >= 0; --i)
+    {
+        Node* child = children[i];
+        if (child->IsTemporary())
+            node->RemoveChild(child);
+    }
 }
 
-PrefabReference::~PrefabReference() = default;
+void PrefabReference::RemoveInstance()
+{
+    if (Node* instanceNode = instanceNode_)
+    {
+        RemoveTemporaryComponents(instanceNode);
+        RemoveTemporaryChildren(instanceNode);
+    }
 
-void PrefabReference::RegisterObject(Context* context)
+    instanceNode_ = nullptr;
+}
+
+void PrefabReference::InstantiatePrefab(const ScenePrefab& nodePrefab)
 {
-    context->AddFactoryReflection<PrefabReference>(Category_Scene);
+    const auto flags = PrefabLoadFlag::KeepExistingComponents | PrefabLoadFlag::KeepExistingChildren
+        | PrefabLoadFlag::LoadAsTemporary | PrefabLoadFlag::IgnoreRootAttributes;
+    PrefabReaderFromMemory reader{nodePrefab};
+    node_->Load(reader, flags);
+}
 
-    URHO3D_ACTION_STATIC_LABEL("Inline", Inline, "Inline the prefab nodes");
+void PrefabReference::CreateInstance(bool tryInplace)
+{
+    // Remove existing instance if moved to another node
+    if (instanceNode_ && instanceNode_ != node_)
+        RemoveInstance();
+
+    // Cannot spawn instance without node
+    if (!node_)
+        return;
 
-    URHO3D_ACCESSOR_ATTRIBUTE("Preserve Transform", GetPreserveTransfrom, SetPreserveTransfrom, bool, false, AM_DEFAULT);
-    URHO3D_MIXED_ACCESSOR_ATTRIBUTE("Prefab", GetPrefabAttr, SetPrefabAttr, ResourceRef, ResourceRef(XMLFile::GetTypeStatic()), AM_DEFAULT);
+    const ScenePrefab& nodePrefab = GetNodePrefab();
+    instanceNode_ = node_;
+    numInstanceComponents_ = nodePrefab.GetComponents().size();
+    numInstanceChildren_ = nodePrefab.GetChildren().size();
+
+    // Try to create inplace first
+    if (tryInplace && TryCreateInplace())
+        return;
+
+    RemoveTemporaryComponents(node_);
```

Comment:
It looks like this would prevent users from having 2 prefab components in the node. Although nothing else would prevent from doing it, so the outcome will be strange behavior and WTF moment. Would be nice if either multiple components work or if it won't be possible to create. Or something along these lines.

Reply:
How would you reload prefab when you have two prefab components in the node?
Like, how should it know which components should be reloaded and which should not?

Reply:
I guess by keeping track of the components? Original implementation had reference to the root prefab node to track it.

Reply:
> I guess by keeping track of the components?

How to track nodes and components that were created indirectly by components from prefab then?
Like AnimatedModel creates bone nodes.

They may be spawned at arbitrary point during prefab lifetime and they should be removed at reload.

Reply:
When you delete AnimatedModel component it deletes the bone nodes.